### PR TITLE
CORTX-30318: Node restart Move current session management to consul

### PR DESCRIPTION
**Pull request by abhijit1patil**
Signed-off-by: abhijit1patil <abhijit.1.patil@seagate.com>

# Problem Statement
- [CORTX-30318 JIRA Ticket](https://jts.seagate.com/browse/CORTX-30318)

# Design
- [ Session Management ](https://seagate-systems.atlassian.net/wiki/spaces/PRIVATECOR/pages/982253682/Session+Management)

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent
# Custom Build and Deployment
- [Custom Build Job](https://eos-jenkins.colo.seagate.com/job/GitHub-custom-ci-builds/job/generic/job/custom-ci/5805/)
- Custom Build Image
cortx-docker.colo.seagate.com/seagate/cortx-all:2.0.0-5805-custom-ci
cortx-docker.colo.seagate.com/seagate/cortx-data:2.0.0-5805-custom-ci
cortx-docker.colo.seagate.com/seagate/cortx-rgw:2.0.0-5805-custom-ci
- [Deployment Job for One node](https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/RGW/job/setup-cortx-rgw-cluster/2526/)
- [Deployment Job for Three node](https://eos-jenkins.colo.seagate.com/job/Cortx-Automation/job/RGW/job/setup-cortx-rgw-cluster/2577/)
- One Node cluster deployed in VM
image

- Three Node Cluster deployed in VM
image

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Dev Testing
testcase0:- To test get() and create()

1. Login with CSM user, save the token 
image

2. crash the pod, once new pod is up.
image

3. Use the same token to hit Any CSM REST API which takes Bearer token 
image

testcase:- To test get_all and delete methods

1. Login with CSM admin user , Create CSM Manage User and Multiple times Login with Manage user credentials save the tokens 
image

2. crash the pod, once new pod is up
image

3. Use the same token to hit Any CSM REST API which takes Bearer token 
image

4. Login with CSM Admin User and delete the  Manage User
 image

5. Now Try to hit CSM API with Manage user session token
image

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: csm/conf/csm_default.conf [hunk 1]
```diff
@@ -88,6 +88,7 @@ CSM:
   username: csm
   password: ''
   password_decryption_key: cortx
+  session_backend: in_memory
```
Comment (rohitkolapkar) [inline csm/conf/csm_default.conf:91]: There should be two possible values for session_backend
1. persistent  -> for consul
2. local -> for in_memory

File: csm/conf/etc/csm/csm.conf [hunk 2]
```diff
@@ -128,6 +128,7 @@ CSM:
     username: "csm"
     password: ""
     password_decryption_key: "cortx"
+    session_backend: "in_memory"
```
Comment (rohitkolapkar) [inline csm/conf/etc/csm/csm.conf:131]: There should be two possible values for session_backend
1. persistent  -> for consul
2. local -> for in_memory

File: csm/conf/csm_default.conf (hunk 1 above)
Comment (abhijit1patil) [reply csm/conf/csm_default.conf:91]: @rohitkolapkar I feel persistent is quite generalize keyword, 
As @pawankumarsrivastava  mentioned in the discussion that we can store in File also or they are might be something new.

File: csm/conf/csm_default.conf [hunk 3]
```diff
@@ -88,6 +88,7 @@ CSM:
   username: csm
   password: ''
   password_decryption_key: cortx
+  session_backend: database
```
Comment (pawankumarsrivastava) [inline csm/conf/csm_default.conf:91]: Please change this to -> CSM>SESSION>storage

The values of storage can be 'local' or 'persistent'. 

You need to map consul to persistent and in-memory to local. 

In future if any new storage comes just map to either of the two

File: csm/conf/etc/csm/csm.conf [hunk 4]
```diff
@@ -128,6 +128,7 @@ CSM:
     username: "csm"
     password: ""
     password_decryption_key: "cortx"
+    session_backend: "database"
```
Comment (pawankumarsrivastava) [inline csm/conf/etc/csm/csm.conf:131]: Please change this to -> CSM>SESSION>storage

The values of storage can be 'local' or 'persistent'. 

You need to map consul to persistent and in-memory to local. 

In future if any new storage comes just map to either of the two

File: csm/conf/csm_default_db.conf [hunk 5]
```diff
@@ -43,3 +43,8 @@ models:
     config:
       consul_db:
         collection: replace_node_data
+  - import_path: csm.core.data.models.session.SessionModel
+    database: consul_db
+    config:
+        consul_db:
+            collection: session_collection
```
Comment (pawankumarsrivastava) [inline csm/conf/csm_default_db.conf:50]: session_collection -> session_mgmt

File: csm/conf/etc/csm/database.conf [hunk 6]
```diff
@@ -50,3 +50,8 @@ models:
     config:
         consul_db:
             collection: "replace_node_data"
+-   import_path: "csm.core.data.models.session.SessionModel"
+    database: "consul_db"
+    config:
+        consul_db:
+            collection: "session_collection"
```
Comment (pawankumarsrivastava) [inline csm/conf/etc/csm/database.conf:57]: session_collection -> session_mgmt

File: csm/core/agent/csm_agent.py [hunk 7]
```diff
@@ -106,7 +106,8 @@ def init():
         auth_service = AuthService()
         user_manager = UserManager(db)
         role_manager = RoleManager(roles)
-        session_manager = SessionManager()
+        session_manager = SessionManager(Conf.get(const.CSM_GLOBAL_INDEX, const.SESSION_BACKEND_KEY),
```
Comment (pawankumarsrivastava) [inline csm/core/agent/csm_agent.py:109]: This should be backward compatible. No need to pass config in constructor. 

Make db parameter as optional .

In the service always have a check for NULL value for db

If you do not get config please keep the default to 'persistent'

File: csm/core/blogic/const.py [hunk 8]
```diff
@@ -945,3 +945,8 @@
 REQUEST_CANCELLED       = 'RequestCancelled'
 SETUP_ERROR             = 'SetupError'
 HTTP_ERROR              = 'HttpError'
+
+# Session Backend factory
+DB                  = 'database'
```
Comment (pawankumarsrivastava) [inline csm/core/blogic/const.py:950]: Please change all these values accordingly

File: csm/core/data/models/session.py [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+from datetime import datetime
+from schematics.types import (StringType, DateTimeType, ModelType,
+                              DictType, ListType)
+
+class PermissionSetModel(CsmModel):
+    ''' Permission Set Model stored in a compact way as a dictionary '''
+
+    _items = DictType(ListType(StringType))
+
+class SessionCredentialsModel(CsmModel):
```
Comment (pawankumarsrivastava) [inline csm/core/data/models/session.py:25]: What value this model class is adding. It only has 1 attribute of string type and no complex data type

File: csm/core/data/models/session.py [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+    @staticmethod
+    def instantiate_session(session_id: str,
+                expiry_time: datetime,
+                credentials: SessionCredentialsModel,
+                permissions: PermissionSetModel):
+        session = SessionModel()
+        session._session_id = session_id
+        session._expiry_time = expiry_time
+        session._credentials = credentials
```
Comment (pawankumarsrivastava) [inline csm/core/data/models/session.py:48]: User id is only the id. Why variable name _credentials is used to store user id?

File: csm/core/data/models/session.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+    """
+    _user_id = StringType()
+
+class SessionModel(CsmModel):
+    """ Session data """
+
+    _id = "_session_id"
+    _session_id = StringType()
+    _expiry_time = DateTimeType()
+    _credentials = ModelType(SessionCredentialsModel)
```
Comment (pawankumarsrivastava) [inline csm/core/data/models/session.py:37]: Is this really required

File: csm/core/services/session_factory.py [hunk 12]
```diff
@@ -0,0 +1,166 @@
...
+from csm.core.blogic import const
+from cortx.utils.data.access import Query
+from cortx.utils.data.access.filters import Compare
+from csm.core.services.permissions import PermissionSet
+from csm.core.blogic.models import CsmModel
+from datetime import datetime
+from csm.core.data.models.session import (SessionModel, SessionCredentialsModel, 
+                                          PermissionSetModel)
+
+class SessionCredentials:
```
Comment (pawankumarsrivastava) [inline csm/core/services/session_factory.py:27]: Rename csm/core/services/session_factory.py --> csm/core/services/session/session_factory.py

**Review by pawankumarsrivastava — CHANGES_REQUESTED** [review]
I have review only half of the PR. Will post more review comments

File: csm/core/data/models/session.py (hunk 10 above)
Comment (abhijit1patil) [reply csm/core/data/models/session.py:48]: I will remove _credentials

File: csm/core/data/models/session.py (hunk 11 above)
Comment (abhijit1patil) [reply csm/core/data/models/session.py:37]: As per the usage of Model .. we have to declare variable as class variable
I will change the data type as per new changes

File: csm/core/data/models/session.py (hunk 9 above)
Comment (abhijit1patil) [reply csm/core/data/models/session.py:25]: I will take directly user id as input in model

File: csm/core/agent/csm_agent.py (hunk 7 above)
Comment (abhijit1patil) [reply csm/core/agent/csm_agent.py:109]: db parameter is optional--> I will add check for null
 I am not passing config in constructor --> I am reading session_backend key word from conf and passing it here
Conf.get(const.CSM_GLOBAL_INDEX, const.SESSION_BACKEND_KEY)

File: csm/conf/etc/csm/database.conf (hunk 6 above)
Comment (abhijit1patil) [reply csm/conf/etc/csm/database.conf:57]: I will update this

File: csm/conf/csm_default_db.conf (hunk 5 above)
Comment (abhijit1patil) [reply csm/conf/csm_default_db.conf:50]: I will update this

File: csm/core/data/models/session.py (hunk 10 above)
Comment (abhijit1patil) [reply csm/core/data/models/session.py:48]: updated

File: csm/core/data/models/session.py (hunk 11 above)
Comment (abhijit1patil) [reply csm/core/data/models/session.py:37]: updated

File: csm/conf/etc/csm/csm.conf (hunk 4 above)
Comment (abhijit1patil) [reply csm/conf/etc/csm/csm.conf]: updated

File: csm/conf/csm_default_db.conf (hunk 5 above)
Comment (abhijit1patil) [reply csm/conf/csm_default_db.conf]: updated

File: csm/conf/etc/csm/database.conf (hunk 6 above)
Comment (abhijit1patil) [reply csm/conf/etc/csm/database.conf]: updated

File: csm/core/agent/csm_agent.py (hunk 7 above)
Comment (abhijit1patil) [reply csm/core/agent/csm_agent.py]: updated

File: csm/core/blogic/const.py (hunk 8 above)
Comment (abhijit1patil) [reply csm/core/blogic/const.py:950]: updated the const

File: csm/core/data/models/session.py (hunk 9 above)
Comment (abhijit1patil) [reply csm/core/data/models/session.py]: Updated

File: csm/core/services/session_factory.py (hunk 12 above)
Comment (abhijit1patil) [reply csm/core/services/session_factory.py]: moved to csm/core/services/session/session_factory.py

File: csm/core/services/session/session_factory.py [hunk 13]
```diff
@@ -0,0 +1,178 @@
...
+        #TODO : Nedd to discuss
+        session_backend_map = {
+            const.LOCAL: const.IN_MEMORY,
+            const.PERSISTENT: const.DB
+        }
+        session_backend = session_backend_map[backend]
+        if session_backend == const.DB:
+            if storage is None:
+                #TODO : Verify the error
+                raise CsmError("Persistent Database is NULL")
```
Comment (rohitkolapkar) [inline csm/core/services/session/session_factory.py:172]: CsmInternalError would be an appropriate error here.

Comment (rohitkolapkar) [inline csm/core/services/session/session_factory.py:172]: CsmInternalError would be an appropriate error here.

File: csm/core/services/session/session_factory.py [hunk 14]
```diff
@@ -0,0 +1,178 @@
...
+class SessionFactory:
+    @staticmethod
+    def get_session(storage: DataBaseProvider=None):
+        backend = Conf.get(const.CSM_GLOBAL_INDEX, const.SESSION_BACKEND_KEY)
+        #TODO : Nedd to discuss
+        session_backend_map = {
+            const.LOCAL: const.IN_MEMORY,
+            const.PERSISTENT: const.DB
+        }
+        session_backend = session_backend_map[backend]
```
Comment (rohitkolapkar) [inline csm/core/services/session/session_factory.py:168]: you can get KeyError if key is not present in map. please handle that.

File: csm/core/services/session/session_factory.py (hunk 13 above)
Comment (abhijit1patil) [reply csm/core/services/session/session_factory.py:172]: ok I will update this

**Review by rohitkolapkar — CHANGES_REQUESTED** [review]

File: csm/core/services/session/session_factory.py (hunk 13 above)
Comment (rohitkolapkar) [reply csm/core/services/session/session_factory.py:172]: please modify the error message too.

Comment (abhijit1patil) [reply csm/core/services/session/session_factory.py:172]: Updated

File: csm/core/services/session/session_factory.py (hunk 14 above)
Comment (abhijit1patil) [reply csm/core/services/session/session_factory.py:168]: updated

File: csm/conf/csm_default.conf (hunk 3 above)
Comment (abhijit1patil) [reply csm/conf/csm_default.conf]: updated

File: csm/core/services/session/session_factory.py (hunk 13 above)
Comment (abhijit1patil) [reply csm/core/services/session/session_factory.py]: updated

File: csm/core/data/models/session.py [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+
+class SessionModel(CsmModel):
+    """ Session Model data """
+
+    _id = "_session_id"
+    _session_id = StringType()
+    _expiry_time = DateTimeType()
+    _user_id = StringType()
+    #TODO: Need to verify and Discuss
+    _permission = DictType(ListType(StringType))
```
Comment (rohitkolapkar) [inline csm/core/data/models/session.py:28]: Hi @abhijit1patil, as we discussed, have you tried this with DictType(SetType(StringType)) ?

**Review by rohitkolapkar — CHANGES_REQUESTED** [review]

File: csm/core/blogic/const.py [hunk 16]
```diff
@@ -952,3 +952,13 @@
...
+IN_MEMORY           = 'in_memory'
+SESSION_STORAGE_KEY = 'CSM>SESSION>storage'
+SESSION_BACKEND_KEY = 'CSM>SESSION>backend'
+LOCAL               = 'local'
+PERSISTENT          = 'persistent'
+IN_MEMORY_DICT      = 'in_memory_dict'
```
Comment (pawankumarsrivastava) [inline csm/core/blogic/const.py:963]: Why this sort of duplicate variables you have taken ? IN_MEMORY_DICT and PERSISTENT_DB

File: csm/core/data/models/session.py [hunk 17]
```diff
@@ -0,0 +1,40 @@
...
+                              DictType, ListType)
+
+class SessionModel(CsmModel):
+    """ Session Model data """
+
+    _id = "_session_id"
+    _session_id = StringType()
+    _expiry_time = DateTimeType()
+    _user_id = StringType()
+    #TODO: Need to verify and Discuss
```
Comment (pawankumarsrivastava) [inline csm/core/data/models/session.py:27]: remove todo

File: csm/core/services/session/session_factory.py [hunk 18]
```diff
@@ -0,0 +1,198 @@
...
+    async def get(self, session_id: Session.Id) -> Optional[Session]:
+        return self._stg.get(session_id, None)
+
+    async def get_all(self):
+        return list(self._stg.values())
+
+    async def store(self, session: Session) -> None:
+        self._stg[session.session_id] = session
+
+class PersistentDatabase:
```
Comment (pawankumarsrivastava) [inline csm/core/services/session/session_factory.py:120]: PersistentDatabase -> Database

File: csm/core/services/session/session_factory.py [hunk 19]
```diff
@@ -0,0 +1,198 @@
...
+
+class SessionFactory:
+    @staticmethod
+    def get_instance(storage: DataBaseProvider=None):
+        # session_backend_keys: Two Level nested map 
+        # First Level Storage :- Local / Persistent
+        # Inner Level(leaf) backend : add here for more session backend option
+        session_backend_keys = {
+            const.LOCAL: {
+                const.IN_MEMORY_DICT:const.IN_MEMORY
```
Comment (pawankumarsrivastava) [inline csm/core/services/session/session_factory.py:180]: Change the variables to a proper name

File: csm/core/services/session/session_factory.py [hunk 20]
```diff
@@ -0,0 +1,198 @@
...
+            },
+            const.PERSISTENT: {
+                const.DB : const.PERSISTENT_DB
+            }
+        }
+        # Leaf level session backend to respective class instance
+        session_backend_instances = {
+            const.PERSISTENT_DB: PersistentDatabase(storage),
```
Comment (pawankumarsrivastava) [inline csm/core/services/session/session_factory.py:188]: Please change variable names

**Review by pawankumarsrivastava — APPROVED** [review]

File: csm/core/services/session/session_factory.py (hunk 18 above)
Comment (abhijit1patil) [reply csm/core/services/session/session_factory.py:120]: updated

File: csm/core/data/models/session.py (hunk 17 above)
Comment (abhijit1patil) [reply csm/core/data/models/session.py:27]: removed

**Review by pawankumarsrivastava — CHANGES_REQUESTED** [review]
Please take care of few minor comments

File: csm/core/blogic/const.py (hunk 16 above)
Comment (abhijit1patil) [reply csm/core/blogic/const.py:963]: removed

File: csm/core/services/session/session_factory.py (hunk 20 above)
Comment (abhijit1patil) [reply csm/core/services/session/session_factory.py:188]: updated

File: csm/core/services/session/session_factory.py (hunk 19 above)
Comment (abhijit1patil) [reply csm/core/services/session/session_factory.py]: updated

**Review by rohitkolapkar — APPROVED** [review]

File: csm/core/services/session_factory.py [hunk 21]
```diff
@@ -0,0 +1,134 @@
...
+class Database:
+    def __init__(self, storage: DataBaseProvider):
+        self.storage = storage
+
+    async def delete(self, session_id: Session.Id) -> None:
+        await self.storage(Session).delete(Compare(Session.Id, '=', session_id))
+
+    async def get(self, session_id: Session.Id) -> Optional[Session]:
+        query = Query().filter_by(Compare(Session.Id, '=', session_id))
+        return list(await self.storage(Session).get(query))
```
[the remark this answers is not in the document]
Comment (abhijit1patil) [reply csm/core/services/session_factory.py]: get() returns list

**Review by pawankumarsrivastava — APPROVED** [review]

**Review by Pranali04796 — APPROVED** [review]

**Pull request merged**